shiny/workflowPlot/server_files/model_data_plots_server.R

```diff
@@ -0,0 +1,193 @@
+# Renders ggplotly 
+
+output$modelDataPlot <- renderPlotly({
+  validate(
+    need(length(input$all_workflow_id) == 1, "Select only ONE workflow ID"),
+    need(length(input$all_run_id) == 1, "Select only ONE run ID"),
+    need(input$load_model > 0, 'Select Load Data'),
+    need(length(input$all_site_id) == 1, 'Select only ONE Site ID'),
+    need(length(input$all_input_id) == 1, 'Select only ONE Input ID'),
+    need(input$load_data > 0, 'Select Load External Data')
+  )
+  plt <- ggplot(data.frame(x = 0, y = 0), aes(x,y)) + 
+    annotate("text", x = 0, y = 0, label = "You are ready to plot!", 
+             size = 10, color = "grey")
+})
+
+# Update units every time a variable is selected
+observeEvent(input$var_name_modeldata, {
+  model.df <- load.model()
+  default.unit <- model.df %>% filter(var_name == input$var_name_modeldata) %>% pull(ylab) %>% unique()
+  updateTextInput(session, "units_modeldata", value = default.unit)
+})
+
+# Check that new units are parsible and can be used for conversion
+observeEvent(input$units_modeldata,{
+  parseable <- ifelse(udunits2::ud.is.parseable(input$units_modeldata), "can", "cannot")
+  if(udunits2::ud.is.parseable(input$units_modeldata)){
+    model.df <- load.model()
+    default.unit <- model.df %>% filter(var_name == input$var_name_modeldata) %>% pull(ylab) %>% unique()
+    if(udunits2::ud.are.convertible(default.unit, input$units_modeldata)){
+      output$unit_text2 <-  renderText({"Conversion possible"})
+    }else{
+      output$unit_text2 <-  renderText({"Units are parsible but conversion is not possible"})
+    }
+  }else{
+    output$unit_text2 <-  renderText({"Units are not parsible, please type units in udunits2 compatible format"})
+  }
+})
```

Comment:
This code appears twice -- once here, and once later on. Might be a good idea to refactor into a `check_units` function that returns the converted units and the message in an `attribute`. The function could live in `PEcAn.utils` or `PEcAn.visualization`. For instance, something like:

```r
convert_units <- function(x, units_modeldata, var_name_modeldata) {
    if (!parseable) return(structure(NA, message = "Units are not parseable"))
    if (parseable && !convertible) return(structure(NA, message = "Parsable but not convertible"))
    out <- udunits2::ud.convert(x, orig_units, out_units)
    structure(out, message = "Units are convertible and have been converted")
}
# ...
observedEvent(input$units_modeldata, { x <- covnert_units(...); renderText({attr(x, "message")})
```

Reply:
I'm going to punt on this until the summer. I think this can also be solved with modularizing, but after taking a crack at it I found it somewhat convoluted. I have my attempts saved for future reference.